### ISSUE TITLE
ModelResult breaks after multiple dumps() loads()

### ISSUE THREAD

**Issue by zobristnicholas** [lmfit/lmfit-py#534]
I'm trying to use the ModelResult.dumps() and ModelResult.loads() classes to serialize the result of a model fit. For my application, this needs to be done more than once.

Unfortunately, calling the dumps()/loads() sequence more than once breaks. It looks like loads() doesn't perfectly recreate the ModelResult.

Here is an example:
```
import lmfit as lm
import numpy as np

def func(x, a):
    return a * x**2

model = lm.Model(func)
params = lm.Parameters()
params.add('a', value=3)

xx = np.linspace(0, 1, 100)
yy = xx**2

result = model.fit(yy, params, x=xx)

result2 = lm.model.ModelResult(model, lm.Parameters())
result2.loads(result.dumps(), funcdefs={'func': func})

result3 = lm.model.ModelResult(model, lm.Parameters())
result3.loads(result2.dumps(), funcdefs={'func': func})  # AttributeError
```
Traceback:
```
AttributeError                            Traceback (most recent call last)
***/test.py in <module>
     18
     19 result3 = lm.model.ModelResult(model, None)
---> 20 result3.loads(result2.dumps(), funcdefs={'func': func})  # AttributeError

***/lmfit/model.py in dumps(self, **kws)
   1598                      'success', 'userargs', 'userkws', 'values', 'var_names',
   1599                      'weights', 'user_options'):
-> 1600             val = getattr(self, attr)
   1601             if isinstance(val, np.bool_):
   1602                 val = bool(val)

***/lmfit/minimizer.py in values(self)
    476     def values(self):
    477         """Return Parameter values in a simple dictionary."""
--> 478         return {name: p.value for name, p in self.result.params.items()}
    479
    480     def __residual(self, fvars, apply_bounds_transformation=True):

AttributeError: 'ModelResult' object has no attribute 'result'
```
Version information:
Python: 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46)
[GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]

lmfit: 0.9.12, scipy: 1.2.0, numpy: 1.15.4, asteval: 0.9.13, uncertainties: 3.0.3, six: 1.12.0

**Comment by newville** [comment]
On Mon, Feb 11, 2019 at 7:00 PM Nicholas Zobrist <notifications@github.com>
wrote:

> I'm trying to use the ModelResult.dumps() and ModelResult.loads() classes
> to serialize the result of a model fit. For my application, this needs to
> be done more than once.
>
> Unfortunately, calling the dumps()/loads() sequence more than once breaks.
> It looks like loads() doesn't perfectly recreate the ModelResult.
>
> Here is an example:
>
> import lmfit as lm
> import numpy as np
>
> def func(x, a):
>     return a * x**2
>
> model = lm.Model(func)
> params = lm.Parameters()
> params.add('a', value=3)
>
> xx = np.linspace(0, 1, 100)
> yy = xx**2
>
> result = model.fit(yy, params, x=xx)
>
> result2 = lm.model.ModelResult(model, lm.Parameters())
> result2.loads(result.dumps(), funcdefs={'func': func})
>
> result3 = lm.model.ModelResult(model, lm.Parameters())
> result3.loads(result2.dumps(), funcdefs={'func': func})  # AttributeError
>
> Traceback:
>
> AttributeError                            Traceback (most recent call last)
> ***/test.py in <module>
>      18
>      19 result3 = lm.model.ModelResult(model, None)
> ---> 20 result3.loads(result2.dumps(), funcdefs={'func': func})  # AttributeError
>
> ***/lmfit/model.py in dumps(self, **kws)
>    1598                      'success', 'userargs', 'userkws', 'values', 'var_names',
>    1599                      'weights', 'user_options'):
> -> 1600             val = getattr(self, attr)
>    1601             if isinstance(val, np.bool_):
>    1602                 val = bool(val)
>
> ***/lmfit/minimizer.py in values(self)
>     476     def values(self):
>     477         """Return Parameter values in a simple dictionary."""
> --> 478         return {name: p.value for name, p in self.result.params.items()}
>     479
>     480     def __residual(self, fvars, apply_bounds_transformation=True):
>
> AttributeError: 'ModelResult' object has no attribute 'result'
>
> Version information:
> Python: 3.6.8 |Anaconda, Inc.| (default, Dec 29 2018, 19:04:46)
> [GCC 4.2.1 Compatible Clang 4.0.1 (tags/RELEASE_401/final)]
>
> lmfit: 0.9.12, scipy: 1.2.0, numpy: 1.15.4, asteval: 0.9.13,
> uncertainties: 3.0.3, six: 1.12.0
>
>
Yeah, it does seem that a loaded ModelResult is not complete in the sense
that it has not been run (or "prepared") so that the `result` attribute has
not been defined.   I'm not entirely sure what the right solution is...

... so I'll come back to your initial statement:

        I'm trying to use the ModelResult.dumps() and ModelResult.loads()
classes to serialize the
        result of a model fit. For my application, this needs to be done
more than once.

And ask: wait, why does saving a "loaded but never run" ModelResult need to
be saved more than once?  I think it won't change unless you re-run the
fit, in which case you might want to save that new result, and you should
be able to do so.

--Matt

**Comment by zobristnicholas** [comment]
I’m working on an application that sends ModelResults to different processes. They must be serialized first. Every once and a while they get sent more than once. It seemed most straightforward to use the built in methods. However, this can also be accomplished by just pickling the result, or only saving to file once and loading the rest of the time like you suggest. 

I started the issue mainly because I was surprised by the error, and it seemed like unintentional behavior. I agree that this is probably an edge case.

**Comment by newville** [comment]
@zobristnicholas Yeah, ModelResult is sort of messy thing.  One could easily interpret that name to mean it holds the entire state of the model and the result, and be ready to redo a fit, make plots and so on.  And it *does* have all this.   But --- expecting to "save" all that and load back to that complete state that can pick up and run in another process is actually asking a lot, especially when we want (and, to be clear, we do) separation of Model definition, data, parameters, and so forth.   

That's not to say we shouldn't try to fix this, just that the fix might be subtle, and I am completely willing to forgive ourselves for this case not being tested.....

**Comment by newville** [comment]
@zobristnicholas this should be fixed in 0.9.13, and we now have a test based on this example.....  I'll close this, but feel free to reopen or raise a related issue if you have trouble with this.
